File: 2021/Day03/Day03.cpp
#include <cassert>
#include <cstdint>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

int64_t bin2Dec(const string& s)
{
    int64_t res = 0;
    int numBits = static_cast<int>(s.size());
    for(int i = 0;i < numBits;++i) {
        res *= 2;
        if(s[i] == '1')
        {
            res += 1;
        }
    }
    return res;
}

static int64_t partA(const vector<string>& allBits)
{
    int numBits = static_cast<int>(allBits.front().size());
    string gamma;
    string episilon;
    for(int i = 0;i < numBits; ++i)
    {
        int numZeros = 0;
        int numOnes = 0;
        for(auto& bits : allBits) 
        {
            if(bits.at(i) == '0')
            {
                numZeros+= 1;
            }
            else 
            {
                assert(bits.at(i) == '1');
                numOnes += 1;
            }
        }
        if(numZeros > numOnes)
        {
            gamma.push_back('0');
            episilon.push_back('1');
        }
        else 
        {
            assert(numZeros != numOnes);
            gamma.push_back('1');
            episilon.push_back('0');
        }
    }
    //cout << gamma << ' ' << bin2Dec(gamma) <<  '\n';
    //cout << episilon << ' ' << bin2Dec(episilon) << '\n';
    return bin2Dec(gamma) * bin2Dec(episilon);
}

vector<string> filterByBit(const vector<string>& allBits, int bitNum, bool mostOrLeast)
{
    int numZeros = 0;
    int numOnes = 0;    
    for(auto& bits : allBits) 
    {
        if(bits.at(bitNum) == '0')
        {
            numZeros+= 1;
        }
        else 
        {
            assert(bits.at(bitNum) == '1');
            numOnes += 1;
        }
    }
    if(numZeros == 0 || numOnes == 0)
    {
        return allBits;
    }
    if(numZeros > numOnes)
    {
        vector<string> retVal;
        if(mostOrLeast)
        {
            for(auto& bits : allBits) 
            {
                if(bits.at(bitNum) == '0') {
                    retVal.push_back(bits);
                }
            }    
        }
        else 
        {
            for(auto& bits : allBits) 
            {
                if(bits.at(bitNum) == '1') {
                    retVal.push_back(bits);
                }
            }
        }
        return retVal;
    }
    else if(numOnes > numZeros)
    {
        vector<string> retVal;
        if(mostOrLeast)
        {
            for(auto& bits : allBits) 
            {
                if(bits.at(bitNum) == '1') {
                    retVal.push_back(bits);
                }
            }    
        }
        else 
        {
            for(auto& bits : allBits) 
            {
                if(bits.at(bitNum) == '0') {
                    retVal.push_back(bits);
                }
            }
        }
        return retVal;
    }
    else 
    {
        vector<string> retVal;
        if(mostOrLeast)
        {
            for(auto& bits : allBits) 
            {
                if(bits.at(bitNum) == '1') {
                    retVal.push_back(bits);
                }
            }    
        }
        else 
        {
            for(auto& bits : allBits) 
            {
                if(bits.at(bitNum) == '0') {
                    retVal.push_back(bits);
                }
            }
        }
        return retVal;
    }
}

static int64_t partB(const vector<string>& allBits)
{
    vector<string> allBitsOxygen(allBits);
    int bitNum = 0;
    while(allBitsOxygen.size() != 1) 
    {
        allBitsOxygen = filterByBit(allBitsOxygen, bitNum, true);
        ++bitNum;
    }

    bitNum = 0;
    vector<string> allBitsCo2(allBits);
    while(allBitsCo2.size() != 1) 
    {
        allBitsCo2 = filterByBit(allBitsCo2, bitNum, false);
        ++bitNum;
    }
    return bin2Dec(allBitsOxygen.front()) * bin2Dec(allBitsCo2.front());
}

int main(int argc,char* argv[]) 
{
    vector<string> allBits;
    string bits;
    while(cin >> bits) 
    {
        allBits.push_back(bits);
    }
    cout << "Part A: " << partA(allBits) << '\n';
    cout << "Part B: " << partB(allBits) << '\n';

    return 0;
}

File: rust/aoc_2021/src/bin/day03.rs
use aoc_2021::*;

fn bin_to_dec(s: &String) -> u64 {
    let mut res = 0;
    for ch in s.chars() {
        res *= 2;
        if ch == '1' {
            res += 1;
        }
    }
    res
}

fn part_a(all_bits: &Vec<String>) -> u64 {
    let n = all_bits[0].len();

    let mut zero_counts = vec![0; n];
    let mut one_counts = vec![0; n];

    for bits in all_bits {
        for (i, ch) in bits.chars().enumerate() {
            if ch == '0' {
                zero_counts[i] += 1;
            } else {
                assert_eq!(ch, '1');
                one_counts[i] += 1;
            }
        }
    }

    let mut gamma = String::new();
    let mut epsilon = String::new();

    for i in 0..n {
        if zero_counts[i] > one_counts[i] {
            gamma.push('0');
            epsilon.push('1');
        } else {
            assert!(zero_counts[i] != one_counts[i]);
            gamma.push('1');
            epsilon.push('0');
        }
    }
    bin_to_dec(&gamma) * bin_to_dec(&epsilon)
}

fn filter_by_bit(all_bits: &Vec<String>, bit_num: usize, most_or_least: bool) -> Vec<String> {
    let mut zero_count = 0;
    let mut one_count = 0;

    for bits in all_bits {
        for ch in bits.chars().skip(bit_num).take(1) {
            if ch == '0' {
                zero_count += 1;
            } else {
                assert_eq!(ch, '1');
                one_count += 1;
            }
        }
    }

    if zero_count == 0 || one_count == 0 {
        return all_bits.clone();
    }
    if zero_count > one_count {
        let mut ret_val: Vec<String> = Vec::new();
        for bits in all_bits {
            for ch in bits.chars().skip(bit_num).take(1) {
                if ch == '0' {
                    if most_or_least {
                        ret_val.push(bits.clone());
                    }
                } else {
                    assert_eq!(ch, '1');
                    if !most_or_least {
                        ret_val.push(bits.clone());
                    }
                }
            }
        }
        return ret_val;
    }

    let mut ret_val: Vec<String> = Vec::new();
    for bits in all_bits {
        for ch in bits.chars().skip(bit_num).take(1) {
            if ch == '1' {
                if most_or_least {
                    ret_val.push(bits.clone());
                }
            } else {
                assert_eq!(ch, '0');
                if !most_or_least {
                    ret_val.push(bits.clone());
                }
            }
        }
    }

    ret_val
}

fn part_b(all_bits: &Vec<String>) -> u64 {
    let mut all_bits_oxygen = all_bits.clone();
    {
        let mut bit_num = 0;
        while all_bits_oxygen.len() != 1 {
            all_bits_oxygen = filter_by_bit(&all_bits_oxygen, bit_num, true);
            bit_num += 1;
        }
    }

    let mut all_bits_co2 = all_bits.clone();
    {
        let mut bit_num = 0;
        while all_bits_co2.len() != 1 {
            all_bits_co2 = filter_by_bit(&all_bits_co2, bit_num, false);
            bit_num += 1;
        }
    }
    bin_to_dec(&all_bits_oxygen[0]) * bin_to_dec(&all_bits_co2[0])
}

fn main() {
    let bits: Vec<String> = read_lines("./day03.txt")
        .unwrap()
        .map(|l| l.unwrap())
        .collect();
    println!("Part A: {}", part_a(&bits));
    println!("Part B: {}", part_b(&bits));
}
